src/Passes/Source/ValidationPass/ValidationPass.cpp

```diff
@@ -0,0 +1,115 @@
+#include <iostream>
+
+namespace microsoft
+{
+namespace quantum
+{
+    llvm::PreservedAnalyses ValidationPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& /*mam*/)
+    {
+
+        for (auto& function : module)
+        {
+            for (auto& block : function)
+            {
+                for (auto& instr : block)
+                {
+                    auto opname = instr.getOpcodeName();
+                    if (opcodes_.find(opname) != opcodes_.end())
+                    {
+                        ++opcodes_[opname];
+                    }
+                    else
+                    {
+                        opcodes_[opname] = 1;
+                    }
+                    auto call_instr = llvm::dyn_cast<llvm::CallBase>(&instr);
+                    if (call_instr != nullptr)
+                    {
+                        auto f = call_instr->getCalledFunction();
+                        if (f == nullptr)
+                        {
+                            continue;
+                        }
+
+                        auto name = static_cast<std::string>(f->getName());
+                        if (f->isDeclaration())
+                        {
+                            if (external_calls_.find(name) != external_calls_.end())
+                            {
+                                ++external_calls_[name];
+                            }
+                            else
+                            {
+                                external_calls_[name] = 1;
+                            }
+                        }
+                        else
+                        {
+                            if (internal_calls_.find(name) != internal_calls_.end())
+                            {
+                                ++internal_calls_[name];
+                            }
+                            else
+                            {
+                                internal_calls_[name] = 1;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        bool raise_exception = false;
+        if (config_.whitelistOpcodes())
+        {
+            auto const& allowed_ops = config_.allowedOpcodes();
+            for (auto const& k : opcodes_)
+            {
+                if (allowed_ops.find(k.first) == allowed_ops.end())
+                {
+                    llvm::errs() << "; Error: '" << k.first << "' is not allowed for this profile.\n";
+                }
+            }
+        }
+
+        auto const& allowed_functions = config_.allowedExternalCallNames();
+        if (config_.whitelistOpcodes())
+        {
+            for (auto const& k : external_calls_)
+            {
+                if (allowed_functions.find(k.first) == allowed_functions.end())
+                {
+                    llvm::errs() << "; Error: '" << k.first << "' is not allowed for this profile.\n";
+                }
+            }
+        }
```

Comment:
The `allowed_functions` (line 83) is only used in subsequent `if` (line 84).
Consider moving line 83 to between the lines 85 and 86, then merging the `if`s in line 84 and 71.

Reply:
Done.